src/contract_response.ts

```diff
@@ -0,0 +1,186 @@
+import { Memo, MemoType, Operation, SorobanRpc, Transaction, xdr } from 'stellar-sdk';
+import { ContractError, ContractErrorType, parseError } from './contract_error.js';
+export interface Result<T, E extends ContractError> {
+  unwrap(): T;
+  unwrapErr(): E;
+  isOk(): boolean;
+  isErr(): boolean;
+}
+
+export class Ok<T, E extends ContractError> implements Result<T, E> {
+  constructor(readonly value: T) {}
+  unwrapErr(): E {
+    throw new Error('No error');
+  }
+  unwrap(): T {
+    return this.value as T;
+  }
+
+  isOk(): boolean {
+    return true;
+  }
+
+  isErr(): boolean {
+    return !this.isOk();
+  }
+}
+
+export class Err<E extends ContractError = ContractError> implements Result<any, E> {
+  constructor(readonly error: E) {}
+  unwrapErr(): E {
+    return this.error;
+  }
+  unwrap(): never {
+    throw new Error(this.error.message);
+  }
+
+  isOk(): boolean {
+    return false;
+  }
+
+  isErr(): boolean {
+    return !this.isOk();
+  }
+}
+export class Resources {
+  fee: number;
+  refundableFee: number;
+  cpuInst: number;
+  readBytes: number;
+  writeBytes: number;
+  readOnlyEntries: number;
+  readWriteEntries: number;
+
+  constructor(
+    fee: number,
+    refundableFee: number,
+    cpuInst: number,
+    readBytes: number,
+    writeBytes: number,
+    readOnlyEntries: number,
+    readWriteEntries: number
+  ) {
+    this.fee = fee;
+    this.refundableFee = refundableFee;
+    this.cpuInst = cpuInst;
+    this.readBytes = readBytes;
+    this.writeBytes = writeBytes;
+    this.readOnlyEntries = readOnlyEntries;
+    this.readWriteEntries = readWriteEntries;
+  }
+
+  /**
+   * Builds a Resources object from TransactionEnvelope
+   * @returns - Resources
+   */
+  static fromTransaction(tx: xdr.TransactionEnvelope | string): Resources {
+    if (typeof tx === 'string') {
+      tx = xdr.TransactionEnvelope.fromXDR(tx, 'base64');
+    }
+    const transaction = tx.v1().tx();
+    const data = transaction.ext().sorobanData();
+    const sorobanResources = data.resources();
+    const footprint = sorobanResources.footprint();
+
+    const fee = transaction.fee();
+    const refundableFee = Number(data.resourceFee().toString());
+    const cpuInst = sorobanResources.instructions();
+    const readBytes = sorobanResources.readBytes();
+    const writeBytes = sorobanResources.writeBytes();
+    const readOnlyEntries = footprint.readOnly().length;
+    const readWriteEntries = footprint.readWrite().length;
+
+    return new Resources(
+      fee,
+      refundableFee,
+      cpuInst,
+      readBytes,
+      writeBytes,
+      readOnlyEntries,
+      readWriteEntries
+    );
+  }
+}
+
+export class ContractResponse<T> {
```

Comment:
any particular reason for the rename?

Reply:
No particular good reason I think while I was doing this I didn't like that it was ContractResult and it contained a result field.